### DISCUSSION TITLE
AutoConfiguration 동작 원리

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#124] [Spring Boot]
.

**Comment by nayonsoso** [comment]
### AutoConfiguration
스프링의 AutoConfiguration은 애플리케이션이 실행될 때 필요한 구성을 자동으로 설정하는 기능이다.
`@SpringBootApplication` 어노테이션에 포함되어있는 `@EnableAutoConfiguration`통해 작동한다.

### 작동 원리
- 스프링 부트는 개발자가 정의한 Component 들을 먼저 스캔한 후, AutoConfiguration의 Bean들을 등록한다.
- AutoConfiguration의 Bean들은 `resource/META-INF/spring.factories `에서 확인할 수 있다.

### 사용 예시
- 스프링 부트는 config 파일 (e.g. application.yml 파일)에 데이터베이스 관련 정보가 정의되어있는 데이터베이스 정보를 통해 데이터베이스 연결에 필요한 DataSource, JdbcTemplate 등을 자동으로 설정한다.
- 스프링 부트는 내장된 웹서버를 자동으로 구성하고, config 파일을 통해 세부 조정을 할 수 있게 한다.

### 커스텀
- 커스텀을 하면 더 폭 넓게 사용할 수 있다.
- https://donghyeon.dev/spring/2020/08/01/스프링부트의-AutoConfiguration의-원리-및-만들어-보기/
- http://dveamer.github.io/backend/SpringBootAutoConfiguration.html

ref. https://tecoble.techcourse.co.kr/post/2021-10-14-springboot-autoconfiguration/

**Comment by gyungchan-jo** [comment]
AutoConfiguration의 시작은 `@SpringBootApplication` 어노테이션 안에 있는`@EnableAutoConfiguration` 이라는 애노테이션입니다.

`@EnableAutoConfiguration`은 `@Import(AutoConfigurationImportSelector.class)`를 통해 자동 구성 클래스를 가져옵니다. 

```java
@Target({ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Inherited
@AutoConfigurationPackage
@Import({AutoConfigurationImportSelector.class})
public @interface EnableAutoConfiguration {
    String ENABLED_OVERRIDE_PROPERTY = "spring.boot.enableautoconfiguration";

    Class<?>[] exclude() default {};

    String[] excludeName() default {};
}
```

자동 구성 클래스를 가져올 때는 AutoConfigurationImportSelector 클래스의 `selectImports(AnnotationMetadata annotationMetadata)` 라는 메서드를 이용하고, `getAutoConfigurationEntry(AnnotationMetadata annotationMetadata);` 메서드를 통해 Import할 클래스가 무엇인지 알 수 있게 됩니다.

- 간단한 메서드 동작 과정 설명 

1. getCandidateConfigurations(annotationMetadata, attributes); - AutoConfiguration의 후보들을 가져온다.
2. removeDuplicates(configurations); - 중복을 제거한다.
3. getExclusions(annotationMetadata, attributes); - 자동 설정에서 제외되는 설정에 대한 정보를 가져온다.
4. configurations.removeAll(exclusions); - 제외되는 설정을 제거한다.
5. getConfigurationClassFilter().filter(configurations); - 필터를 적용한다.

[참고 블로그](https://velog.io/@realsy/Spring-Boot-AutoConfiguration-%EB%8F%99%EC%9E%91-%EC%9B%90%EB%A6%AC)
